database/sqlite/sqlite_functions.c

```diff
@@ -231,170 +228,20 @@ int sql_database_size(void)
-            freez(item->name);
-            freez(item);
-            break;
-        }
-        uuid_cache = &(item)->next;
-    }
-    return uuid;
-}
-
-int cache_host_charts(RRDHOST *host)
-{
-#ifdef ENABLE_CACHE_CHARTS
-    int rc;
-    sqlite3_stmt *res = NULL;
-
-    if (!db_meta) {
-        error_report("Database has not been initialized");
-        return 0;
-    }
-
-    int count = 0;
-
-    if (!res) {
-        rc = sqlite3_prepare_v2(
-            db_meta, "select chart_id, type, id, name from chart where host_id = @host order by rowid desc;", -1, &res, 0);
-        if (rc != SQLITE_OK)
-            return 0;
-    }
-
-    rc = sqlite3_bind_blob(res, 1, &host->host_uuid, sizeof(host->host_uuid), SQLITE_STATIC);
-    if (rc != SQLITE_OK) {
-        error_report("Failed to bind host_uuid to find host charts");
-        goto done;
-    }
-    while (sqlite3_step(res) == SQLITE_ROW) {
-        add_in_uuid_cache(
-            &host->uuid_cache, (uuid_t *)sqlite3_column_blob(res, 0), (const char *)sqlite3_column_text(res, 1),
-            (const char *)sqlite3_column_text(res, 2), (const char *)sqlite3_column_text(res, 3));
-        count++;
-    }
-
-done:
-    rc = sqlite3_finalize(res);
-    if (unlikely(rc != SQLITE_OK))
-        error_report("Failed to finalize statement when loading host charts, rc = %d", rc);
-    return count;
-#else
-    return 0;
-#endif
-}
-
-int cache_chart_dimensions(RRDSET *st)
-{
-#ifdef ENABLE_CACHE_DIMENSIONS
-    sqlite3_stmt *res = NULL;
-    int rc;
-
-    if (!db_meta) {
-        error_report("Database has not been initialized");
-        return 0;
-    }
-
-    int count = 0;
-
-    if (!res) {
-        rc = sqlite3_prepare_v2(db_meta, "select dim_id, id, name from dimension where chart_id = @chart order by rowid desc;", -1, &res, 0);
-        if (rc != SQLITE_OK) {
-            error_report("Failed to prepare statement to find chart dimensions");
-            return 0;
-        }
-    }
-
-    rc = sqlite3_bind_blob(res, 1, st->chart_uuid, sizeof(*st->chart_uuid), SQLITE_STATIC);
-    if (rc != SQLITE_OK) {
-        error_report("Failed to bind chart_uuid to find chart dimensions");
-        goto done;
-    }
-
-    while (sqlite3_step(res) == SQLITE_ROW) {
-        add_in_uuid_cache(
-            &st->state->uuid_cache, (uuid_t *)sqlite3_column_blob(res, 0), NULL,
-            (const char *)sqlite3_column_text(res, 1), (const char *)sqlite3_column_text(res, 2));
-        count++;
-    }
-done:
-    rc = sqlite3_finalize(res);
-    if (unlikely(rc != SQLITE_OK))
-        error_report("Failed to finalize statement when loading chart dimensions, rc = %d", rc);
-    return count;
-#else
-    return 0;
-#endif
-}
-
 #define FIND_UUID_TYPE  "select 1 from host where host_id = @uuid union select 2 from chart where chart_id = @uuid union select 3 from dimension where dim_id = @uuid;"
 
 int find_uuid_type(uuid_t *uuid)
 {
-    sqlite3_stmt *res = NULL;
+    static __thread sqlite3_stmt *res = NULL;
```

Comment:
is this not GCC specific?

Reply:
Additionally if you use ptherads api: `In addition pthread_key_create can optionally accept a destructor function that will automatically be called at thread exit, if the thread-specific data is not NULL`

Reply:
The pthread thing has CPU overhead because it does allocations and a lookup. We already use thread local storage in netdata in the core:
https://github.com/netdata/netdata/blob/3f8e44d1a6c501feeeacab3b3bf86ec0e401d4a3/libnetdata/threads/threads.c#L19

Reply:
Well there are multiple options here. Question is how much we care to clean up as per previous discussion.

Reply:
My main argument here is that there is no problem with GCC since this kind of code already exists in netdata, so we shouldn't worry about compatibility.

Reply:
Yes that is certainly valid argument. I see 2 minor problems with this (don't worry won't block this PR on this though):
1. cleanup
2. portability

Is it worth killing both? How much is the performance impact?

Reply:
For this specific function it may not matter because it will be used only during migration, but in general preparing the statement helps a lot.

Adding some quick logging to check

Preparing every time
```
2020-11-23 12:04:02: netdata INFO  : MAIN : SQLITE: FIND UUID 49 usec
2020-11-23 12:04:02: netdata INFO  : MAIN : SQLITE: FIND UUID 31 usec
2020-11-23 12:04:02: netdata INFO  : MAIN : SQLITE: FIND UUID 26 usec
2020-11-23 12:04:02: netdata INFO  : MAIN : SQLITE: FIND UUID 24 usec
2020-11-23 12:04:02: netdata INFO  : MAIN : SQLITE: FIND UUID 23 usec
2020-11-23 12:04:02: netdata INFO  : MAIN : SQLITE: FIND UUID 23 usec
2020-11-23 12:04:02: netdata INFO  : MAIN : SQLITE: FIND UUID 23 usec
2020-11-23 12:04:02: netdata INFO  : MAIN : SQLITE: FIND UUID 23 usec
2020-11-23 12:04:02: netdata INFO  : MAIN : SQLITE: FIND UUID 23 usec
2020-11-23 12:04:02: netdata INFO  : MAIN : SQLITE: FIND UUID 23 usec
```

Preparing once
```
2020-11-23 12:02:16: netdata INFO  : MAIN : SQLITE: FIND UUID 43 usec
2020-11-23 12:02:16: netdata INFO  : MAIN : SQLITE: FIND UUID 13 usec
2020-11-23 12:02:16: netdata INFO  : MAIN : SQLITE: FIND UUID 7 usec
2020-11-23 12:02:16: netdata INFO  : MAIN : SQLITE: FIND UUID 6 usec
2020-11-23 12:02:16: netdata INFO  : MAIN : SQLITE: FIND UUID 6 usec
2020-11-23 12:02:16: netdata INFO  : MAIN : SQLITE: FIND UUID 6 usec
2020-11-23 12:02:16: netdata INFO  : MAIN : SQLITE: FIND UUID 5 usec
2020-11-23 12:02:16: netdata INFO  : MAIN : SQLITE: FIND UUID 5 usec
2020-11-23 12:02:16: netdata INFO  : MAIN : SQLITE: FIND UUID 6 usec
2020-11-23 12:02:16: netdata INFO  : MAIN : SQLITE: FIND UUID 6 usec
```

We can refine the code in future PRs to properly release the memory allocated by having a generic prepare function that will keep track of the statements and release during shutdown.

Reply:
We can also resolve the memory leak by pushing in the pthread cleanup callback stack with via [pthread_cleanup_push()](https://man7.org/linux/man-pages/man3/pthread_cleanup_push.3.html).